File: android/app/src/main/java/com/ezequielbrrt/domemory/services/seasons/Season.kt
package com.ezequielbrrt.domemory.services.seasons

import com.ezequielbrrt.domemory.core.time.DayKey
import com.ezequielbrrt.domemory.services.levels.LevelCurve
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

/**
 * One seasonal level progression decoded from a child of `/seasons` (spec 9.2).
 *
 * Validation fails **closed** on structure (a season missing [enabled], a
 * `levelCount` under 1, or an emoji pool with fewer than [LevelCurve.maxPairs]
 * distinct entries is rejected outright — see [SeasonDecoder.decode]) and falls
 * **back** on decoration: [accentColor] and the artwork URLs are validated at
 * decode time and simply resolve to `null` when malformed, so a typo in the
 * console degrades the surface to a flat colour instead of costing the player
 * the whole season (spec 9.3).
 */
data class Season(
    val id: String,
    val enabled: Boolean,
    val startDay: String?,
    val endDay: String?,
    val priority: Int,
    val levelCount: Int,
    val icon: String,
    /** Validated `#RRGGBB` (8-digit forms are not supported), or null to fall back to brand primary. */
    val accentColor: String?,
    /** Validated absolute `https` URL, or null to fall back to a flat colour. */
    val backgroundImageURL: String?,
    /** Dark-appearance replacement for [backgroundImageURL]; null means one image serves both. */
    val backgroundImageURLDark: String?,
    /** Artwork for the menu card; null falls back to [accentColor]. */
    val cardImageURL: String?,
    val emojiPool: List<String>,
    /** Locale key (normalized `_`→`-`, lowercased) to title. */
    val title: Map<String, String>,
    /** Locale key to subtitle; a locale present in [title] but missing a subtitle reads as `""`. */
    val subtitle: Map<String, String>,
) {
    fun isActive(day: String) = enabled && startDay != null && endDay != null && day in startDay..endDay

    /** All non-null artwork URLs, for prefetching (spec 9.7): card, then background light, then dark. */
    fun artworkUrls(): List<String> = listOfNotNull(cardImageURL, backgroundImageURL, backgroundImageURLDark)

    /**
     * Whole days from [today] (a `YYYYMMDD` day key) to the season's last day, inclusive —
     * 0 means the season ends today. Null when [endDay] failed to parse.
     */
    fun daysRemaining(today: String): Int? {
        val end = endDay ?: return null
        return runCatching {
            val endDate = LocalDate.parse(end, DAY_KEY_FORMAT)
            val todayDate = LocalDate.parse(today, DAY_KEY_FORMAT)
            ChronoUnit.DAYS.between(todayDate, endDate).toInt().coerceAtLeast(0)
        }.getOrNull()
    }

    private companion object {
        val DAY_KEY_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")
    }
}

object SeasonDecoder {
    fun decode(raw: String): List<Season> = runCatching {
        val root = JSONObject(raw)
        root.keys().asSequence().mapNotNull { id -> root.optJSONObject(id)?.let { decodeOne(id, it) } }.toList()
    }.getOrDefault(emptyList())

    /** Highest `priority` among active seasons; ties break on the smallest id (spec 9.4). */
    fun active(seasons: List<Season>, day: String): Season? = seasons.filter { it.isActive(day) }
        .sortedWith(compareByDescending<Season> { it.priority }.thenBy { it.id })
        .firstOrNull()

    private fun decodeOne(id: String, json: JSONObject): Season? {
        val count = json.optInt("levelCount", 0); if (count < 1) return null
        // Order-preserving dedup: the seeded shuffle depends on pool order (root CLAUDE.md).
        val pool = json.optJSONArray("emojiPool").strings().map(String::trim).filter(String::isNotEmpty).distinct()
        if (pool.size < LevelCurve.maxPairs) return null
        val start = day(json.optString("startDate")); val end = day(json.optString("endDate"))
        val strings = json.optJSONObject("strings")
        val titles = mutableMapOf<String, String>(); val subtitles = mutableMapOf<String, String>()
        strings?.keys()?.forEach { key ->
            strings.optJSONObject(key)?.let { value ->
                titles[key.normalize()] = value.optString("title")
                subtitles[key.normalize()] = value.optString("subtitle")
            }
        }
        return Season(
            id = id,
            enabled = json.optBoolean("enabled", false),
            startDay = start,
            endDay = end,
            priority = json.optInt("priority", 0),
            levelCount = count,
            icon = json.optString("icon").ifBlank { "✨" },
            // optString("...") reads back "" when the key is absent, and every parser below
            // treats a blank string as "not provided" — so absent and blank are, correctly,
            // indistinguishable here.
            accentColor = parseAccentColor(json.optString("accentColor")),
            backgroundImageURL = parseArtworkUrl(json.optString("backgroundImageURL")),
            backgroundImageURLDark = parseArtworkUrl(json.optString("backgroundImageURLDark")),
            cardImageURL = parseArtworkUrl(json.optString("cardImageURL")),
            emojiPool = pool,
            title = titles,
            subtitle = subtitles,
        )
    }

    private fun day(value: String): String? = DayKey.parse(value)

    private fun String.normalize() = replace('_', '-').lowercase()

    private fun JSONArray?.strings(): List<String> =
        this?.let { (0 until it.length()).mapNotNull { i -> it.optString(i).takeIf(String::isNotBlank) } }.orEmpty()

    /**
     * `#RRGGBB` (with or without the leading `#`) to a normalized `#RRGGBB` string.
     * Null for anything else, including the 8-digit form, which this payload does not define.
     */
    fun parseAccentColor(raw: String?): String? {
        var value = raw?.trim() ?: return null
        if (value.isEmpty()) return null
        if (value.startsWith("#")) value = value.substring(1)
        if (value.length != 6 || value.any { !it.isHexDigit() }) return null
        return "#" + value.uppercase()
    }

    /**
     * `https` URL with a non-empty host, or null for anything else (blank, malformed, or
     * cleartext `http` — ATS/cleartext-blocking parity, spec 9.3).
     */
    fun parseArtworkUrl(raw: String?): String? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null
        if (uri.scheme?.lowercase() != "https") return null
        if (uri.host.isNullOrEmpty()) return null
        return trimmed
    }

    private fun Char.isHexDigit(): Boolean = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
}

/**
 * Resolves a season's `strings` map for a device locale (spec 9.5).
 *
 * The lookup is built from progressively **shorter** prefixes of the locale identifier,
 * always ending in `en` — `es-419` → `["es-419", "es", "en"]`, `zh-Hans-CN` →
 * `["zh-hans-cn", "zh-hans", "zh", "en"]`. First hit wins; no hit returns null so the
 * caller can apply its own app-side fallback (spec: "app-side fallback title").
 *
 * This was a real shipped iOS bug: the lookup only ever *shortened* the identifier, so a
 * key was only reachable from the exact identifier that produced it — `es-419` was
 * unreachable from `es_MX`, and `zh-Hans` was unreachable from any mainland device
 * because iOS canonicalizes `zh_Hans_CN` to `zh_CN`, dropping the script. The fix lived in
 * the *catalog* (publish both `zh-Hans` and `zh-CN`, both `es-419` and `es`) rather than in
 * the algorithm, which this port keeps as-is.
 *
 * **This does not translate directly to Android.** `Locale.toLanguageTag()` reports
 * different identifiers than iOS's `Locale.identifier` — notably `zh-Hans-CN` (Android
 * keeps the script subtag; iOS drops it), `es-419` and `pt-BR`. See
 * `SeasonLocaleResolutionTest` for the pinned, real `Locale.toLanguageTag()` outputs this
 * assumes.
 */
object SeasonLocaleResolver {
    /** Author-written keys are matched case-insensitively, `_` and `-` both accepted. */
    fun normalizeKey(key: String): String = key.replace('_', '-').lowercase()

    /** The ordered lookup candidates for a locale identifier, always ending in `en`. */
    fun candidates(languageTag: String): List<String> {
        val components = normalizeKey(languageTag).split('-').filter { it.isNotEmpty() }
        val result = mutableListOf<String>()
        var index = components.size
        while (index > 0) {
            result += components.subList(0, index).joinToString("-")
            index--
        }
        if ("en" !in result) result += "en"
        return result
    }

    /**
     * First candidate present in [values], else null. [values]' own keys are normalized
     * defensively — `Season`'s decoder already normalizes `strings` keys at decode time, but
     * this stays tolerant of a caller (or test) passing author-written keys as-is, matching
     * "author-written keys are matched case-insensitively" above.
     */
    fun resolve(values: Map<String, String>, languageTag: String): String? {
        val normalized = values.mapKeys { (key, _) -> normalizeKey(key) }
        for (candidate in candidates(languageTag)) {
            normalized[candidate]?.let { return it }
        }
        return null
    }

    /** Title/subtitle pair for [languageTag], or null when no candidate (including `en`) hits. */
    fun resolveText(season: Season, languageTag: String): Pair<String, String>? {
        val title = resolve(season.title, languageTag) ?: return null
        val subtitle = resolve(season.subtitle, languageTag) ?: ""
        return title to subtitle
    }
}

File: ios/DoMemory/DoMemory/Services/Seasons/Season.swift
//
//  Season.swift
//  DoMemory
//
//  Firebase-supplied seasonal level progression ("Spooky Season",
//  "Christmas Season"). Firebase supplies the emoji pool, the length and the
//  theming — never the individual boards, which are generated locally and
//  deterministically by SeasonProgressService.
//

import Foundation

/// One seasonal level progression, decoded from a child of the Realtime
/// Database `/seasons` node.
///
/// `/seasons` is a *dictionary* keyed by season id (unlike `/data`, which is an
/// array), so `id` does not appear in the value body. `SeasonCatalogService`
/// injects the dictionary key as `id` before decoding, which also lets the
/// decoded value round-trip cleanly through the UserDefaults cache.
struct Season: Codable, Identifiable, Hashable {
    /// The title/subtitle pair for a single locale.
    struct LocalizedText: Codable, Hashable {
        let title: String
        let subtitle: String

        init(title: String, subtitle: String) {
            self.title = title
            self.subtitle = subtitle
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            title = try container.decode(String.self, forKey: .title)
            subtitle = try container.decodeIfPresent(String.self, forKey: .subtitle) ?? ""
        }
    }

    /// The Realtime Database key of the season, e.g. `spooky-2026`.
    let id: String
    /// Kill switch. Absent means `false`: a season only appears when Firebase
    /// says so explicitly.
    let enabled: Bool
    /// Inclusive first day, `YYYY-MM-DD`.
    let startDate: String
    /// Inclusive last day, `YYYY-MM-DD`.
    let endDate: String
    /// Highest priority wins when several seasons are active on the same day.
    let priority: Int
    /// How many levels the season holds. A season is finite.
    let levelCount: Int
    /// Optional display glyph for the menu card. Phase 3 owns the fallback.
    let icon: String?
    /// Optional accent colour as a `#RRGGBB` hex string. Phase 3 owns parsing
    /// and the fallback.
    let accentColor: String?
    /// Optional full-bleed artwork for the season's level map, as an absolute
    /// `https` URL.
    ///
    /// Remote rather than bundled because a season is published from Firebase
    /// without an app update: art shipped in the binary would only ever cover
    /// the seasons that existed at build time, and every player who had not
    /// updated would see the *previous* season's picture behind the new
    /// season's levels. `Season+Presentation` owns validation and the fallback.
    let backgroundImageURL: String?
    /// Optional dark-appearance replacement for `backgroundImageURL`.
    ///
    /// One image cannot serve both appearances: the level tiles are near-white
    /// in light mode and near-black in dark, so artwork that separates from the
    /// tiles in one collapses into them in the other. Absent means the season
    /// has a single image and both appearances use it.
    let backgroundImageURLDark: String?
    /// Optional artwork for the season's menu card, under the same contract as
    /// `backgroundImageURL`. The card is the season's accent colour with white
    /// text in both appearances, so it takes one image, not a pair.
    let cardImageURL: String?
    /// Emoji the season's boards are dealt from. Deduplicated at decode time;
    /// see `minimumEmojiPoolSize`.
    let emojiPool: [String]
    /// Locale code (`en`, `es-419`, …) to display strings.
    let strings: [String: LocalizedText]

    // MARK: - Pool validation

    /// The largest board a season will ever be asked to deal.
    ///
    /// `LevelCurve.pairAnchors` tops out at 12 pairs
    /// (`Services/Levels/LevelCurve.swift:15`) and holds the last anchor's
    /// value for every level beyond it, so asking the curve for an arbitrarily
    /// high level yields that cap without duplicating the number here — if the
    /// curve is ever retuned, this follows it.
    ///
    /// A season whose pool is smaller than this cannot fill its own late
    /// boards, so decoding rejects it outright rather than dealing a short
    /// board at level 25.
    ///
    /// **Cross-reference — keep in sync.** `firebase/scripts/upload_seasons.py` restates
    /// this rule as `MINIMUM_EMOJI_POOL_SIZE` so a season is rejected before it
    /// is published rather than skipped after. Python cannot import Swift, so
    /// that script hardcodes 12 and will *not* follow a retuned
    /// `LevelCurve.pairAnchors` the way this constant does. Change both
    /// together; the script carries the matching pointer back here.
    static let minimumEmojiPoolSize = LevelCurve.pairs(for: .max)

    /// Title shown when a season carries no usable strings for the current
    /// locale and no `en` entry either.
    ///
    /// `Strings` is a plain `Foundation`-only enum of `NSLocalizedString`
    /// lookups, which the services layer already reaches directly
    /// (`PurchaseService`, `SeasonProgressService`), so a model reading it
    /// pulls in no UI dependency.
    static let fallbackTitle = Strings.seasonFallbackTitle

    // MARK: - Decoding

    init(
        id: String,
        enabled: Bool,
        startDate: String,
        endDate: String,
        priority: Int,
        levelCount: Int,
        icon: String?,
        accentColor: String?,
        backgroundImageURL: String? = nil,
        backgroundImageURLDark: String? = nil,
        cardImageURL: String? = nil,
        emojiPool: [String],
        strings: [String: LocalizedText]
    ) {
        self.id = id
        self.enabled = enabled
        self.startDate = startDate
        self.endDate = endDate
        self.priority = priority
        self.levelCount = levelCount
        self.icon = icon
        self.accentColor = accentColor
        self.backgroundImageURL = backgroundImageURL
        self.backgroundImageURLDark = backgroundImageURLDark
        self.cardImageURL = cardImageURL
        self.emojiPool = emojiPool
        self.strings = strings
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        id = try container.decode(String.self, forKey: .id)
        // Fail closed on every optional field: a season missing its kill switch
        // is treated as switched off rather than silently shipped.
        enabled = try container.decodeIfPresent(Bool.self, forKey: .enabled) ?? false
        startDate = try container.decode(String.self, forKey: .startDate)
        endDate = try container.decode(String.self, forKey: .endDate)
        priority = try container.decodeIfPresent(Int.self, forKey: .priority) ?? 0
        levelCount = try container.decode(Int.self, forKey: .levelCount)
        icon = try container.decodeIfPresent(String.self, forKey: .icon)
        accentColor = try container.decodeIfPresent(String.self, forKey: .accentColor)
        // Artwork is decoration. A blank or malformed URL resolves to nil in
        // `Season+Presentation` and the surface falls back to its flat colour,
        // so a typo in the console never costs the player the season itself.
        backgroundImageURL = try container.decodeIfPresent(String.self, forKey: .backgroundImageURL)
        backgroundImageURLDark = try container.decodeIfPresent(String.self, forKey: .backgroundImageURLDark)
        cardImageURL = try container.decodeIfPresent(String.self, forKey: .cardImageURL)
        strings = try container.decodeIfPresent([String: LocalizedText].self, forKey: .strings) ?? [:]

        guard levelCount >= 1 else {
            throw DecodingError.dataCorruptedError(
                forKey: .levelCount,
                in: container,
                debugDescription: "Season \"\(id)\" has levelCount \(levelCount); a season needs at least one level."
            )
        }

        // A duplicated emoji would deal two identical pairs — four matching
        // cards — which the matching rules cannot resolve, so the pool is
        // deduplicated (order preserved, so the seeded shuffle stays stable)
        // before it is measured.
        let rawPool = try container.decode([String].self, forKey: .emojiPool)
        var seen = Set<String>()
        let pool = rawPool.filter { emoji in
            guard !emoji.isEmpty else { return false }
            return seen.insert(emoji).inserted
        }
        guard pool.count >= Season.minimumEmojiPoolSize else {
            throw DecodingError.dataCorruptedError(
                forKey: .emojiPool,
                in: container,
                debugDescription: """
                    Season "\(id)" has \(pool.count) distinct emoji; \
                    \(Season.minimumEmojiPoolSize) are required to fill its late boards.
                    """
            )
        }
        emojiPool = pool
    }

    // MARK: - Locale resolution

    /// Display strings for `locale`, resolved through
    /// `es-419` → `es` → `en` → an app-side fallback.
    ///
    /// The chain is built from progressively shorter prefixes of the locale
    /// identifier rather than just "exact, then language", so a script-bearing
    /// identifier such as `zh_Hans_CN` still reaches a `zh-Hans` entry instead
    /// of falling straight through to English.
    func text(for locale: Locale = .current) -> LocalizedText {
        Season.resolveText(from: strings, identifier: locale.identifier)
    }

    static func resolveText(from strings: [String: LocalizedText], identifier: String) -> LocalizedText {
        // Author-written keys are matched case-insensitively so `pt-BR` and
        // `pt-br` behave the same in the console.
        var lookup: [String: LocalizedText] = [:]
        for (key, value) in strings {
            lookup[normalizedLocaleKey(key)] = value
        }
        for candidate in localeCandidates(for: identifier) where lookup[candidate] != nil {
            return lookup[candidate]!
        }
        return LocalizedText(title: fallbackTitle, subtitle: "")
    }

    /// The ordered lookup keys tried for a locale identifier, always ending in
    /// `en`. `es_419` yields `["es-419", "es", "en"]`.
    static func localeCandidates(for identifier: String) -> [String] {
        let components = normalizedLocaleKey(identifier)
            .split(separator: "-")
            .map(String.init)
            .filter { !$0.isEmpty }

        var candidates: [String] = []
        var index = components.count
        while index > 0 {
            candidates.append(components.prefix(index).joined(separator: "-"))
            index -= 1
        }
        if !candidates.contains("en") {
            candidates.append("en")
        }
        return candidates
    }

    /// Apple identifiers use `_` separators (`es_419`), Firebase keys use `-`
    /// (`es-419`). Normalise both to lowercase hyphenated form.
    private static func normalizedLocaleKey(_ identifier: String) -> String {
        identifier.replacingOccurrences(of: "_", with: "-").lowercased()
    }

    // MARK: - Activation window

    /// `enabled`, and `date` falls inside `[startDate, endDate]` inclusive.
    ///
    /// The window is evaluated as *wall-clock days in the player's own
    /// calendar*, matching `DailyChallengeService`, which derives its day from
    /// `Calendar.current` through `dailyChallengeSeed(for:calendar:)`. Both
    /// therefore roll over together at the player's local midnight, so a
    /// player never sees the daily challenge advance to a new day while the
    /// season card still believes it is yesterday.
    ///
    /// The comparison is done on the same zero-padded `YYYYMMDD` day key rather
    /// than on parsed `Date` values: string ordering on that format is
    /// chronological, the bounds are inclusive by construction, and no
    /// timezone offset can shift a boundary day by one.
    func isActive(on date: Date = Date(), calendar: Calendar = .current) -> Bool {
        guard enabled else { return false }
        guard let start = Season.dayKey(fromISODay: startDate),
              let end = Season.dayKey(fromISODay: endDate) else {
            // An unparseable window fails closed.
            return false
        }
        let today = dailyChallengeSeed(for: date, calendar: calendar)
        return today >= start && today <= end
    }

    /// `"2026-10-01"` → `"20261001"`. Returns nil for anything that is not a
    /// strict `YYYY-MM-DD`.
    static func dayKey(fromISODay day: String) -> String? {
        let parts = day.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3,
              parts[0].count == 4, parts[1].count == 2, parts[2].count == 2,
              parts.allSatisfy({ $0.allSatisfy(\.isASCII) && $0.allSatisfy(\.isNumber) }) else {
            return nil
        }
        return parts.joined()
    }
}
